{Pylint} Fix use-dict-literal

atombrella opened the pull request:
**Related command**

Various commands are "affected" by this. Please check the diff.

**Description**

[flake8-comprehensions](https://pypi.org/project/flake8-comprehensions/) has the two suggestions I try to remedy.

```
    Rewrite set([]) as set()
    Rewrite dict([]) as {}
```

You can check the marginal benefits with `timeit` in `ipython`.

I wonder why these two rules are disabled in `pylintrc`. Maybe it's for historic reasons.

```
    use-dict-literal,
    consider-using-dict-items,
```

- [X] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [X] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [X] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

bebound commented:
Thanks for your contribution.

The disabled rules are introduced in https://github.com/Azure/azure-cli/pull/20192. I think it's okay to fix them. Could you please also remove them from `pylintrc`?

atombrella commented:
> Thanks for your contribution.
> 
> The disabled rules were introduced in #20192. I think it's okay to fix them. Could you please also remove them from `pylintrc`?

I can do that, but it'll be quite a bit of work. You unfortunately still accept PRs with `dict()`.

jiasli commented:
`use-dict-literal`: https://pylint.readthedocs.io/en/latest/user_guide/messages/refactor/use-dict-literal.html
`consider-using-dict-items`: https://pylint.pycqa.org/en/latest/user_guide/messages/convention/consider-using-dict-items.html

> The disabled rules were introduced in #20192.

https://github.com/Azure/azure-cli/pull/20192 was created by me. I ignored them simply because we didn't have bandwidth to work on those issues.

So yes, please also remove them from

https://github.com/Azure/azure-cli/blob/be4af714ff08cf76741ecb468b890a72dfaf9f58/pylintrc#L45-L46

jiasli commented:
I will be fixing `consider-using-dict-items` in https://github.com/Azure/azure-cli/pull/30341.

jiasli commented:
@atombrella, could you remove 

https://github.com/Azure/azure-cli/blob/be4af714ff08cf76741ecb468b890a72dfaf9f58/pylintrc#L45

jiasli commented:
More errors are discovered after disabling `use-dict-literal`:

```
2024-11-20T09:02:22.7856436Z ERROR: ************* Module azure.cli.command_modules.feedback.custom
2024-11-20T09:02:22.7857058Z src/azure-cli/azure/cli/command_modules/feedback/custom.py:503:19: R1735: Consider using '{"name_len": 0, "success_len": 0, "time_len": 0}' instead of a call to 'dict'. (use-dict-literal)
2024-11-20T09:02:22.7857614Z ************* Module azure.cli.command_modules.containerapp._utils
2024-11-20T09:02:22.7858237Z src/azure-cli/azure/cli/command_modules/containerapp/_utils.py:1977:22: R1735: Consider using '{"fullTag": tag, "version": SemVer.parse(re_matches[0][0]), "framework": tag_split[2], ... }' instead of a call to 'dict'. (use-dict-literal)
2024-11-20T09:02:22.7858824Z ************* Module azure.cli.command_modules.appservice.custom
2024-11-20T09:02:22.7859395Z src/azure-cli/azure/cli/command_modules/appservice/custom.py:1004:13: R1735: Consider using '{"resource_group_name": resource_group_name, "name": name, "site_envelope": instance, ... }' instead of a call to 'dict'. (use-dict-literal)
2024-11-20T09:02:22.7874325Z src/azure-cli/azure/cli/command_modules/appservice/custom.py:1106:13: R1735: Consider using '{"resource_group_name": resource_group_name, "name": name, "site_envelope": instance, ... }' instead of a call to 'dict'. (use-dict-literal)
2024-11-20T09:02:22.7874608Z ************* Module azure.cli.command_modules.vm.custom
2024-11-20T09:02:22.7874902Z src/azure-cli/azure/cli/command_modules/vm/custom.py:4220:18: R1735: Consider using '{}' instead of a call to 'dict'. (use-dict-literal)
2024-11-20T09:02:22.7875272Z ************* Module azure.cli.command_modules.sql._params
2024-11-20T09:02:22.7875670Z src/azure-cli/azure/cli/command_modules/sql/_params.py:92:36: R1735: Consider using '{"B": 1, "kB": 1024, "MB": 1024 * 1024, "GB": 1024 * 1024 * 1024, "TB": 1024 * 1024 * 1024 * 1024, ... }' instead of a call to 'dict'. (use-dict-literal)
2024-11-20T09:02:22.7876200Z src/azure-cli/azure/cli/command_modules/sql/_params.py:295:63: R1735: Consider using '{"B": 1.0 / (1024 * 1024 * 1024), "kB": 1.0 / (1024 * 1024), "MB": 1.0 / 1024, ... }' instead of a call to 'dict'. (use-dict-literal)
2024-11-20T09:02:22.7876881Z ************* Module azure.cli.core.commands.template_create
2024-11-20T09:02:22.7877348Z src/azure-cli-core/azure/cli/core/commands/template_create.py:62:28: R1735: Consider using '{"name": parent_value, "resource_group": resource_group_name, "namespace": parent_type.split('/')[0], ... }' instead of a call to 'dict'. (use-dict-literal)
2024-11-20T09:02:22.7878020Z src/azure-cli-core/azure/cli/core/commands/template_create.py:72:28: R1735: Consider using '{"name": property_value, "resource_group": resource_group_name, "namespace": property_type.split('/')[0], ... }' instead of a call to 'dict'. (use-dict-literal)
```

@atombrella, could you help fix them too?

atombrella commented:
> More errors are discovered after disabling `use-dict-literal`:
> 
> ```
> 2024-11-20T09:02:22.7856436Z ERROR: ************* Module azure.cli.command_modules.feedback.custom
> 2024-11-20T09:02:22.7857058Z src/azure-cli/azure/cli/command_modules/feedback/custom.py:503:19: R1735: Consider using '{"name_len": 0, "success_len": 0, "time_len": 0}' instead of a call to 'dict'. (use-dict-literal)
> 2024-11-20T09:02:22.7857614Z ************* Module azure.cli.command_modules.containerapp._utils
> 2024-11-20T09:02:22.7858237Z src/azure-cli/azure/cli/command_modules/containerapp/_utils.py:1977:22: R1735: Consider using '{"fullTag": tag, "version": SemVer.parse(re_matches[0][0]), "framework": tag_split[2], ... }' instead of a call to 'dict'. (use-dict-literal)
> 2024-11-20T09:02:22.7858824Z ************* Module azure.cli.command_modules.appservice.custom
> 2024-11-20T09:02:22.7859395Z src/azure-cli/azure/cli/command_modules/appservice/custom.py:1004:13: R1735: Consider using '{"resource_group_name": resource_group_name, "name": name, "site_envelope": instance, ... }' instead of a call to 'dict'. (use-dict-literal)
> 2024-11-20T09:02:22.7874325Z src/azure-cli/azure/cli/command_modules/appservice/custom.py:1106:13: R1735: Consider using '{"resource_group_name": resource_group_name, "name": name, "site_envelope": instance, ... }' instead of a call to 'dict'. (use-dict-literal)
> 2024-11-20T09:02:22.7874608Z ************* Module azure.cli.command_modules.vm.custom
> 2024-11-20T09:02:22.7874902Z src/azure-cli/azure/cli/command_modules/vm/custom.py:4220:18: R1735: Consider using '{}' instead of a call to 'dict'. (use-dict-literal)
> 2024-11-20T09:02:22.7875272Z ************* Module azure.cli.command_modules.sql._params
> 2024-11-20T09:02:22.7875670Z src/azure-cli/azure/cli/command_modules/sql/_params.py:92:36: R1735: Consider using '{"B": 1, "kB": 1024, "MB": 1024 * 1024, "GB": 1024 * 1024 * 1024, "TB": 1024 * 1024 * 1024 * 1024, ... }' instead of a call to 'dict'. (use-dict-literal)
> 2024-11-20T09:02:22.7876200Z src/azure-cli/azure/cli/command_modules/sql/_params.py:295:63: R1735: Consider using '{"B": 1.0 / (1024 * 1024 * 1024), "kB": 1.0 / (1024 * 1024), "MB": 1.0 / 1024, ... }' instead of a call to 'dict'. (use-dict-literal)
> 2024-11-20T09:02:22.7876881Z ************* Module azure.cli.core.commands.template_create
> 2024-11-20T09:02:22.7877348Z src/azure-cli-core/azure/cli/core/commands/template_create.py:62:28: R1735: Consider using '{"name": parent_value, "resource_group": resource_group_name, "namespace": parent_type.split('/')[0], ... }' instead of a call to 'dict'. (use-dict-literal)
> 2024-11-20T09:02:22.7878020Z src/azure-cli-core/azure/cli/core/commands/template_create.py:72:28: R1735: Consider using '{"name": property_value, "resource_group": resource_group_name, "namespace": property_type.split('/')[0], ... }' instead of a call to 'dict'. (use-dict-literal)
> ```
> 
> @atombrella, could you help fix them too?

Yes. I'll try to allocate some time this week to fix them :)

atombrella commented:
I've fixed them. There weren't that many after all :) There are quite of cases like:

```python
        if self.__loadbalancer_models is None:
            load_balancer_models = {}
            load_balancer_models["ManagedClusterLoadBalancerProfile"] = self.ManagedClusterLoadBalancerProfile
            load_balancer_models[
                "ManagedClusterLoadBalancerProfileManagedOutboundIPs"
            ] = self.ManagedClusterLoadBalancerProfileManagedOutboundIPs
            load_balancer_models[
                "ManagedClusterLoadBalancerProfileOutboundIPs"
            ] = self.ManagedClusterLoadBalancerProfileOutboundIPs
            load_balancer_models[
                "ManagedClusterLoadBalancerProfileOutboundIPPrefixes"
            ] = self.ManagedClusterLoadBalancerProfileOutboundIPPrefixes
            load_balancer_models["ResourceReference"] = self.ResourceReference
            self.__loadbalancer_models = SimpleNamespace(**load_balancer_models)
        return self.__loadbalancer_models
```

If you use PyCharm, it provides a quick and easy way to directly create the dictionary in a single-step.

https://pypi.org/project/flake8-comprehensions/ has lots of valuable suggestions. I've fixed the `dict.fromkeys` in a couple of cases, as well as the list comprehensions.

bebound commented:
> More errors are discovered after disabling `use-dict-literal`:

Some `use-dict-literal` issue can be fixed with `ruff check ./src --select C408 --unsafe-fixes`

Ref: https://github.com/astral-sh/ruff/issues/970#issuecomment-2490038018

atombrella commented:
> > More errors are discovered after disabling `use-dict-literal`:
> 
> Some `use-dict-literal` issue can be fixed with `ruff check ./src --select C408 --unsafe-fixes`, but C408 may also fix other linter issue.
> 
> Ref: [astral-sh/ruff#970 (comment)](https://github.com/astral-sh/ruff/issues/970#issuecomment-2490038018)

I did check with ruff to find all of them :) I hadn't heard of ruff before, it seems to quite useful/powerful. C408 is indeed a superset of https://pylint.readthedocs.io/en/stable/user_guide/messages/refactor/use-dict-literal.html

There are many great suggestions from flake8-comprehensions. For instance, I experimented briefly today with one of the examples from https://www.geeksforgeeks.org/python-dictionary-fromkeys-method/ to check the performance improvements of using `dict.fromkeys` instead of `{x: None for x in iterable}` It's about ~15%. I fixed the single case that this flake8 plugin found, as well as a few of the list comprehensions. 

Note that in some places, I reformatted the code slightly. IMO it improves readability to split comprehensions over more lines. One case is:

```python
info_dict = {
    lun: {"managedDisk": {'storageAccountType': None}}
    for lun in dummy_expected if lun != "os"
}
```

atombrella commented:
@bebound Could you help with the failed "Azure.azure-cli (Codegen Coverage)". I don't think my changes are related.

wangzelin007 commented:
/azp run

wangzelin007 commented:
I created a [PR](https://github.com/Azure/azure-cli/pull/30413), but I couldn't reproduce the issue. I'll try running `/azp run` to see if it resolves the problem.

atombrella commented:
> I created a [PR](https://github.com/Azure/azure-cli/pull/30413), but I couldn't reproduce the issue. I'll try running `/azp run` to see if it resolves the issue.

Seems to be ok again :) Everything is green.

jiasli commented on src/azure-cli-testsdk/azure/cli/testsdk/scenario_tests/tests/test_recording_processor.py:
```diff
@@ -107,11 +107,14 @@ def test_subscription_recording_processor_for_response(self):
 
         for template in body_templates:
             mock_sub_id = str(uuid.uuid4())
-            mock_response = dict({'body': {}})
```
This `dict()` is a no-op, not sure why it was written like that in the first place.

jiasli commented:
@atombrella, Thank you very much for the contribution!

atombrella commented:
> @atombrella, Thank you very much for the contribution!

@jiasli You're welcome :smiley: Note that since my PR was approved and merged, something was merged that now violates the fixed linting rule.

```
Running pylint on modules...
Pylint: PASSED
ERROR: ************* Module azure.cli.command_modules.vm.azure_stack.custom
src/azure-cli/azure/cli/command_modules/vm/azure_stack/custom.py:4250:18: R1735: Consider using '{}' instead of a call to 'dict'. (use-dict-literal)
```

jiasli commented:
> Note that since my PR was approved and merged, something was merged that now violates the fixed linting rule.

Fixed by https://github.com/Azure/azure-cli/pull/30442
